src/deepqmc/physics.py

```diff
@@ -46,6 +47,224 @@ def electronic_potential(rs):
+
+
+@jax.vmap
+def sph2cart(sph, r=1):
+    # This function transforms from spherical to cartesian coordinates.
+    theta = sph[0]
+    phi = sph[1]
+    rsin_theta = r * jnp.sin(theta)
+    x = rsin_theta * jnp.cos(phi)
+    y = rsin_theta * jnp.sin(phi)
+    z = r * jnp.cos(theta)
+    return jnp.array([x, y, z])
+
+
+@jax.jit
+def get_unit_icosahedron_sph():
+    # Basic definition of unit icosahedron vertices in spherical coordinates.
+    unit_icosahedron_sph = []
+    unit_icosahedron_sph.append([0, 0])
+    unit_icosahedron_sph.append([jnp.pi, 0])
+    for j in range(5):
+        unit_icosahedron_sph.append([jnp.arctan(2), jnp.pi / 5 * 2 * j])
+        unit_icosahedron_sph.append([jnp.pi - jnp.arctan(2), jnp.pi / 5 * (2 * j - 1)])
+    return jnp.array(unit_icosahedron_sph)
+
+
+# UNIT_ICOSAHEDRON is (12,3) array of unit icosahedron vertices
+UNIT_ICOSAHEDRON = sph2cart(get_unit_icosahedron_sph())
+QUADRATURE_THETAS = get_unit_icosahedron_sph()[:, 0]
+
+
+def get_quadrature_points(nucleus_position, rs):
+    """
+    Transform electron configuration 'rs' of size (N,3) into an array quadrature points.
+
+    Return array of size (N,12,N,3) that includes all the 12 quadrature point
+    configurations (i.e. reference electron position is shifted to another icosahedron
+    vertex) corresponding to N different reference electron.
+    """
+
+    norm = jnp.linalg.norm(rs - nucleus_position, axis=-1)
+    theta = jnp.arccos((rs - nucleus_position)[..., 2] / norm)
+    phi = jnp.arctan2((rs - nucleus_position)[..., 1], (rs - nucleus_position)[..., 0])
+
+    # rotation matrix about y-axis by angle theta
+    y_rot = jnp.array(
+        [
+            [jnp.cos(theta), jnp.zeros_like(theta), jnp.sin(theta)],
+            [jnp.zeros_like(theta), jnp.ones_like(theta), jnp.zeros_like(theta)],
+            [-jnp.sin(theta), jnp.zeros_like(theta), jnp.cos(theta)],
+        ]
+    )
+    y_rot = jnp.moveaxis(y_rot, -1, -3)  # shape: (3,3,num_e) -> (num_e,3,3)
+
+    # rotation matrix about z-axis by angle phi
+    z_rot = jnp.array(
+        [
+            [jnp.cos(phi), -jnp.sin(phi), jnp.zeros_like(phi)],
+            [jnp.sin(phi), jnp.cos(phi), jnp.zeros_like(phi)],
+            [jnp.zeros_like(phi), jnp.zeros_like(phi), jnp.ones_like(phi)],
+        ]
+    )
+    z_rot = jnp.moveaxis(z_rot, -1, -3)  # shape: (3,3,num_e) -> (num_e,3,3)
+
+    # auxiliary function applying the rotation and translation to be vmapped
+    def transform_coordinates(norm, z_rot, y_rot, r, nucleus_position):
+        return norm * (z_rot @ y_rot @ r) + nucleus_position
+
+    # vmapping to include N different rotations corresponding to each electron position
+    transform_coordinates = jax.vmap(
+        transform_coordinates, in_axes=(-1, -3, -3, None, None)
+    )
+    # vmapping to be able to transform all 12 icosahedron points at the same time
+    transform_coordinates = jax.vmap(
+        transform_coordinates, in_axes=(None, None, None, -2, None)
+    )
```

Comment:
This is because I need to vectorize the function along two extra dimensions. One dimension corresponds to the electron index and the other dimension covers the 12 icosahedron points. I believe this is not possible to be done with a single `vmap`, and I can't think of a more elegant way to write it